compiler/gradient_with_order.cc

```diff
@@ -243,4 +245,192 @@ void AddGradientNodesForTrainingWithOrders(Graph* graph, const std::vector<Order
     graph->ResetGradients();
 }
 
+void AddGradInputs(Graph* fwd_graph, Graph* bwd_graph) {
+    for (Value* value : fwd_graph->output_values()) {
+        Value* grad = bwd_graph->AddInputValue("grad_in@" + value->name(), value->type());
+        value->set_grad(grad);
+    }
+}
+
+void AddRetainedParts(Graph* fwd_graph, Graph* bwd_graph, std::map<Value*, Value*>* staged) {
+    for (auto& p : *staged) {
+        Value* value = p.first;
+
+        GraphBuilder gbf(fwd_graph, "retain", value);
+        GraphBuilder gbb(bwd_graph, "retain", value);
+        const std::string& name = "retained_" + value->name();
+
+        Value* o = fwd_graph->AddOutputValue(name, value->type());
+        gbf.Op(Node::kIdentity, {value}, o);
+
+        Value* i = bwd_graph->AddInputValue(name, value->type());
+
+        // Update the staged value to the retained one
+        p.second = i;
+    }
+}
+
+void AddGradientNodesForTrainingWithOrders(Graph* fwd_graph, Graph* bwd_graph, const std::vector<Order>& orders) {
```

Comment:
I think this function is very similar to the single graph mode. Cannot we just call this version from the other version by `AddGradientNodesForTrainingWithOrders(graph, graph, orders)`? I think, it should just work with a few `fwd_graph == bwd_graph` check in theory?

Reply:
I'm not sure if it is possible, but I will try. At least, I think we need to avoid calling `AddRetainedParts` when `fwd_graph == bwd_graph`. I don't know if I'm missing other things.

Reply:
I merged the two versions of `AddGradientNodesForTrainingWithOrders`. It seems working.

Reply:
Thanks!